system_setup/server/controllers/configure.py

```diff
@@ -40,89 +41,189 @@ def __init__(self, app):
-            locGenNeedsWrite = n == 1
-
-        if locGenNeedsWrite:
-            try:
-                with open(localeGenPath, "wt") as f:
-                    f.write(fileContents)
-            except IOError as err:
-                log.error("Failed to modify %s file. %s", localeGenPath, err)
-
-    async def _install_check_lang_support_packages(self, lang, env):
+        try:
+            # Altering locale-gen script to output to the desired folder.
+            with open(cmdFile, "r+") as f:
+                script = f.read()
+                pattern = r'(localedef .+) (\|\| :; \\)'
+                localeDefRe = re.compile(pattern)
+                replacement = r'\1 "{}/" \2'.format(outDir)
+                (fileContents, n) = localeDefRe.subn(replacement, script,
+                                                     count=1)
+                if n != 1:
+                    log.error("locale-gen script contents were unexpected."
+                              " Aborting mock creation")
+                    return ("", False)
+
+                f.seek(0)
+                f.write(fileContents)
+
+            return (cmdFile, True)
+
+        except IOError as err:
+            log.error("Failed to modify %s file. %s", cmdFile, err)
+            return ("", False)
+
+    def __update_locale_cmd(self, lang) -> list[str]:
+        """ Adds mocking cli to update-locale if in dry-run. """
+        updateLocCmd = ["update-locale", "LANG={}".format(lang)]
+        if self.app.opts.dry_run:
+            defaultLocDir = os.path.join(self.model.root,
+                                         "etc/default/")
+            os.makedirs(defaultLocDir, exist_ok=True)
+            updateLocCmd += ["--locale-file",
+                             os.path.join(defaultLocDir, "locale"),
+                             "--no-checks"]
+
+        return updateLocCmd
+
+    async def _activate_locale(self, lang, env) -> bool:
+        """ Last commands to run for locale support. Returns True on success"""
+
+        (locGenCmd, ok) = self.__locale_gen_cmd()
+        if ok is False:
+            log.error("Locale generation failed.")
+            return False
+
+        updateCmd = self.__update_locale_cmd(lang)
+        cmds = [[locGenCmd], updateCmd]
+        for cmd in cmds:
+            cp = await arun_command(cmd, env=env)
+            if cp.returncode != 0:
+                log.error('Command "{}" failed with return code {}'
+                          .format(cp.args, cp.returncode))
+                return False
+        return True
+
+    async def _install_check_lang_support_packages(self, lang, env) -> bool:
         """ Installs any packages recommended by
             check-language-support command. """
 
         clsCommand = "check-language-support"
+        # lang may have more than 5 chars and be separated by dot or space.
+        clsLang = lang.split('@')[0].split('.')[0].split(' ')[0]
```

Comment:
I would trust you on this one :)

Reply:
Just for the record: :smile:

![image](https://user-images.githubusercontent.com/11138291/140921773-889a115e-d12e-4abd-8586-d36f5a8bcc72.png)

Reply:
maybe a regexp would be more appropriate. Something along the lines:

```
langs = ['nan_tw@latin utf-8', 'en_US', 'C.UTF8' ]

r = re.compile(r'([^.@]+)', re.IGNORECASE)

for l in langs:
      m = r.match(l)
      print(m.groups())
 
('nan_tw',)
('en_US',)
('C',)

```

Plus some error handling in case there is no match of course.